### PR TITLE
   * [tosa] Support for AtenNativeLayerNormOp

### PR DESCRIPTION
**Pull request by anupgangwar**
* [tosa] Support for AtenPermuteOp

Signed-off-by: Anup Gangwar <anup.gangwar@arm.com>

### CODE REVIEW

**Comment by anupgangwar** [comment]
```
Summary:
    Passed: 72
    Expectedly Failed: 196
```
Tagging @cathyzhyi for review.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -1892,26 +1919,184 @@ LogicalResult ConvertAtenOp<AtenBatchNormOp>::matchAndRewrite(
...
+    while (static_cast<int64_t>(toReduceShape.size()) != meanAndVarShapeRank) {
+      toReduceShape.back() = 1;
+      sumDiv =
+          rewriter
+              .create<tosa::ReduceSumOp>(
+                  op.getLoc(),
+                  RankedTensorType::get(toReduceShape,
+                                        inputType.getElementType()),
+                  sumDiv, rewriter.getI64IntegerAttr(toReduceShape.size() - 1))
+              .getResult();
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2007]: nit: all operations with `::mlir::OpTrait::OneResult` trait doesn't need `getResult()`. This is the same for reshape and a few other ops.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -1892,26 +1919,184 @@ LogicalResult ConvertAtenOp<AtenBatchNormOp>::matchAndRewrite(
...
+      toReduceShape.pop_back();
+    }
+
+    return rewriter
+        .create<tosa::ReshapeOp>(op.getLoc(), outType, sumDiv,
+                                 rewriter.getI64ArrayAttr(outShape))
+        .getResult();
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2014]: nit: `getResult` is not needed.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -1892,26 +1919,184 @@ LogicalResult ConvertAtenOp<AtenBatchNormOp>::matchAndRewrite(
...
+                                  {static_cast<float>(elemCnt)}, {1})
+          .getValue();
+  Value elemCntRcp = rewriter
+                         .create<tosa::ReciprocalOp>(
+                             op.getLoc(), elemCntConst.getType(), elemCntConst)
+                         .getResult();
+
+  // Broadcast type and shape for various intermediate values.
+  SmallVector<int64_t> bcastOutShape;
+  for (auto en : llvm::enumerate(inputType.getShape()))
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2034]: nit: {} is needed for for block with multiple lines.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -1892,26 +1919,184 @@ LogicalResult ConvertAtenOp<AtenBatchNormOp>::matchAndRewrite(
...
+      computeSumAndReshape(squareSum, inputType, bcastOutType, bcastOutShape);
+  Value varianceVal =
+      rewriter
+          .create<tosa::MulOp>(op.getLoc(), bcastOutType, squareSumReduced,
+                               elemCntRcp, /*shift=*/0)
+          .getResult();
+
+  // Reshape weight and bias.
+  SmallVector<int64_t> weightAndBiasBcastShape;
+  for (auto en : llvm::enumerate(inputType.getShape()))
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2068]: nit: need {}

File: e2e_testing/torchscript/batchnorm.py [hunk 5]
```diff
@@ -97,7 +97,7 @@ def __init__(self):
     @export
     @annotate_args([
         None,
-        ([2, 5, 2, 2, 3], torch.float32, True),
+        ([5, 2, 2, 3], torch.float32, True),
```
Comment (cathyzhyi) [inline e2e_testing/torchscript/batchnorm.py:100]: Is it possible to keep the original test but add one specifically for TOSA?

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: e2e_testing/torchscript/batchnorm.py (hunk 5 above)
Comment (anupgangwar) [reply e2e_testing/torchscript/batchnorm.py:100]: Added one with '4D' in the name.

**Comment by anupgangwar** [comment]
@cathyzhyi If the changes look good, do merge the PR as well since @sjarus is out.

**Pull request merged**